posts/inside-rust/2023-05-03-stabilizing-async-fn-in-trait.md

```diff
@@ -0,0 +1,229 @@
+
+```rust
+async fn do_health_check_par(hc: impl HealthCheck) {
+    tokio::task::spawn(async move {
+        if !hc.check().await {
+            log_health_check_failure().await;
+        }
+    });
+}
+```
+
+The problem is that the future returned by `hc.check()` isn't guaranteed to be `Send`. It might access non-Send data. The solution is to add a `Send` bound, but given that this is an async function, it's not obvious how to do that. How do we talk about the future returned by a call to `hc.check()`? Associated return types provide the answer. We can convert the above function to use an explicit type parameter `HC` (instead of `impl HealthCheck`) and then add a new bound, `HC::check(): Send`. This says "the value returned by `HC::check` must be of `Send` type":
+
+```rust
+async fn do_health_check_par<HC>(hc: HC)
+where
+    HC: HealthCheck + Send + 'static,
+    HC::check(): Send, // <-- associated return type
+{
+    tokio::task::spawn(async move {
+        if !hc.check().await {
+            log_health_check_failure().await;
+        }
+    });
+}
+```
+
+Of course, it's kind of unfortunate that we had to rewrite from taking an `impl HealthCheck` to an explicit `HC` type parameter in order to use this notation. [RFC 2289](https://github.com/rust-lang/rfcs/pull/2289), "associated type bounds", introduced a compact notation to address this problem. That RFC is not part of this MVP, but if it were stabilized, then one could simply write:
+
+```rust
+async fn do_health_check_par(hc: impl HealthCheck<check(): Send> + Send + 'static) {
+    //                                            -------------
+    tokio::task::spawn(async move {
+        if !hc.check().await {
+            log_health_check_failure().await;
+        }
+    });
+}
+```
+
+In our [previous post][pp], we [hypothesized](https://blog.rust-lang.org/inside-rust/2022/11/17/async-fn-in-trait-nightly.html#hypothesis-this-is-uncommon) that this problem might not occur often in practice. However, our case studies found that it comes up quite frequently, and so we decided that a solution is needed. We explored a number of solutions and concluded that associated return types (ARTs) are a flexible and reasonably ergonomic building block, which makes them a great fit for an MVP.
+ 
+**Status:** Associated return types have an experimental implementation and we are currently drafting an RFC. There are several open bugs that will need to be fixed. We also found that in traits with many methods, ARTs become verbose, and in the future may consider more concise syntax (see below).
+
+[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=2066934a05cb9eafc0b47af7bdf8c57f)
+
+## MVP Part 3: "impl trait in traits" (return position)
```

Comment:
This feels separable enough from AFIT that we could stabilize them separately, even though they obviously share a lot of the underlying implementation. In my mind they're separate features rather than different facets of the same one. That said, I'm in favor of stabilizing both and if it's easier to bundle them together then that's fine.

Reply:
I think the AFIT story feels incomplete without RPITIT, but RPITIT can stand alone without AFIT. We could reorder the sections to reflect this, but since the post is focused on AFIT, maybe the current ordering makes the most sense.

Reply:
I agree that this seems like the best ordering!

Reply:
I added some wording to the status line to reinforce that RPITIT can stand alone.